### ISSUE TITLE
KLDiv vs CrossEntropy in attention loss

### ISSUE THREAD

**Issue by Stasiche** [espnet/espnet#4358]
Hello! Could you please, explain me is there any strong reason to use KLdiv loss here instead of classic (in classification tasks) Crossentropy loss?
https://github.com/espnet/espnet/blob/39bae01e4a132da69b9b0d025da8c579a5f38b77/espnet2/asr/espnet_model.py#L139

**Comment by pyf98** [comment]
```
KLDiv(true, pred) = CrossEntropy(true, pred) - Entropy(true)
```

The entropy of the true distribution will be a constant w.r.t. the network parameters, regardless of label smoothing. So, I think we can use either.

Regarding the implementation, I think it’s because PyTorch `CrossEntropyLoss` couldn’t support non-one-hot target distribution which is required for label smoothing. `KLDivLoss` supports it.

I just checked the latest PyTorch documentation. Starting from PyTorch 1.10, `torch.nn.CrossEntropyLoss` directly supports label smoothing: 
https://pytorch.org/docs/1.10/generated/torch.nn.CrossEntropyLoss.html?highlight=crossentropy#torch.nn.CrossEntropyLoss

But for PyTorch<=1.9.1, it is not supported: https://pytorch.org/docs/1.9.1/generated/torch.nn.CrossEntropyLoss.html?highlight=crossentropy#torch.nn.CrossEntropyLoss

**Comment by Stasiche** [comment]
> 

Okay, thank you very much!
